### ISSUE TITLE
StackOverflow Error

### ISSUE THREAD

**Issue by davidzentlermunro** [JuliaStats/GLM.jl#210]
having happily used GLM for a while, I started getting an unhelpful "StackOverflow" Error when running this code:

`data_reg=readtable("C:\\Onedrive\\KORV with frictions\\Data\\KORV_Frictions_Reg.csv", header=true)
data=DataFrame(data_reg)`
`regh_u=glm(@formula(L_U ~  K_EQ_CR + K_EQ_LAG + K_STR_CR + K_STR_LAG + REL_P_EQ_LAG + IND_BUS_CYC_LAG + YEAR), data, Normal(), IdentityLink())`

I get error message:

> Error evaluating KORVfrictions_3stage_byyear_withsmoothing.jl
LoadError: StackOverflowError:
while loading C:\Onedrive\KORV with frictions\Code\KORVfrictions_3stage_byyear_withsmoothing.jl, in expression starting on line 226

I've attached relevant file (had to save it xlsx as not allowed to attach a csv file for some reason)  - had been working fine a few months ago. 

[KORV_Frictions_Reg.xlsx](https://github.com/JuliaStats/GLM.jl/files/1619263/KORV_Frictions_Reg.xlsx)

**Comment by nalimilan** [comment]
Confirmed here with latest master. There's also a deprecation warning from DataFrames, probably not related but worth checking:
```julia
WARNING: calling vector of vectors constructor without passing column names is deprecated
Stacktrace:
 [1] depwarn(::String, ::Symbol) at ./deprecated.jl:70
 [2] DataFrames.DataFrame(::Array{Array{T,1} where T,1}) at /home/milan/.julia/DataFrames/src/deprecated.jl:4
 [3] #ModelFrame#22(::Dict{Any,Any}, ::Type{T} where T, ::StatsModels.Terms, ::DataFrames.DataFrame) at /home/milan/.julia/StatsModels/src/modelframe.jl:144
 [4] (::Core.#kw#Type)(::Array{Any,1}, ::Type{StatsModels.ModelFrame}, ::StatsModels.Terms, ::DataFrames.DataFrame) at ./<missing>:0
 [5] #ModelFrame#27(::Array{Any,1}, ::Type{T} where T, ::StatsModels.Formula, ::DataFrames.DataFrame) at /home/milan/.julia/StatsModels/src/modelframe.jl:156
 [6] (::Core.#kw#Type)(::Array{Any,1}, ::Type{StatsModels.ModelFrame}, ::StatsModels.Formula, ::DataFrames.DataFrame) at ./<missing>:0
 [7] #fit#53(::Dict{Any,Any}, ::Array{Any,1}, ::Function, ::Type{GLM.GeneralizedLinearModel}, ::StatsModels.Formula, ::DataFrames.DataFrame, ::Distributions.Normal{Float64}, ::Vararg{Any,N} where N) at /home/milan/.julia/StatsModels/src/statsmodel.jl:53
 [8] fit(::Type{GLM.GeneralizedLinearModel}, ::StatsModels.Formula, ::DataFrames.DataFrame, ::Distributions.Normal{Float64}, ::GLM.IdentityLink) at /home/milan/.julia/StatsModels/src/statsmodel.jl:53
 [9] #glm#17(::Array{Any,1}, ::Function, ::StatsModels.Formula, ::DataFrames.DataFrame, ::Distributions.Normal{Float64}, ::Vararg{Any,N} where N) at /home/milan/.julia/GLM/src/glmfit.jl:288
 [10] glm(::StatsModels.Formula, ::DataFrames.DataFrame, ::Distributions.Normal{Float64}, ::Vararg{Any,N} where N) at /home/milan/.julia/GLM/src/glmfit.jl:288
 [11] macro expansion at ./REPL.jl:97 [inlined]
 [12] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:73
while loading no file, in expression starting on line 0
ERROR: StackOverflowError:
Stacktrace:
 [1] #fit#16(::Array{Any,1}, ::Function, ::Type{GLM.GeneralizedLinearModel}, ::Array{Float64,2}, ::Array{Union{Float64, Missings.Missing},1}, ::Distributions.Normal{Float64}, ::GLM.IdentityLink) at /home/milan/.julia/GLM/src/glmfit.jl:285 (repeats 79999 times)
```

**Comment by dmbates** [comment]
Actually the deprecation warning is in `StatsModels`.  I'll take a look.

I haven't looked in detail at why this causes a `StackOverFlowError` but the root cause is fitting a `LinearModel` using the iterative method for a `GeneralizedLinearModel`.

This model can be fit as
```julia
julia> fit(LinearModel, @formula(L_U ~ K_EQ_CR + K_EQ_LAG + K_STR_CR + K_STR_LAG + REL_P_EQ_LAG + IND_BUS_CYC_LAG + YEAR), data)
StatsModels.DataFrameRegressionModel{GLM.LinearModel{GLM.LmResp{Array{Float64,1}},GLM.DensePredChol{Float64,Base.LinAlg.Cholesky{Float64,Array{Float64,2}}}},Array{Float64,2}}

Formula: L_U ~ 1 + K_EQ_CR + K_EQ_LAG + K_STR_CR + K_STR_LAG + REL_P_EQ_LAG + IND_BUS_CYC_LAG + YEAR

Coefficients:
                    Estimate   Std.Error   t value Pr(>|t|)
(Intercept)         -23.2199     20.2237  -1.14815   0.2606
K_EQ_CR          -5.59005e-5  6.51018e-5 -0.858664   0.3978
K_EQ_LAG          2.33942e-5  6.56897e-5  0.356132   0.7244
K_STR_CR         -0.00263883 0.000547447  -4.82026    <1e-4
K_STR_LAG         0.00149792 0.000582174   2.57298   0.0157
REL_P_EQ_LAG       -0.374453    0.232684  -1.60928   0.1188
IND_BUS_CYC_LAG    0.0220212   0.0265041  0.830861   0.4131
YEAR               0.0123778   0.0106946   1.15739   0.2569
```

We may want to consider detecting cases of fitting a `GeneralizedLinearModel` with the `Normal` distribution and the `IdentityLink` and redirecting it to a call to `lm` or to `fit(LinearModel, ...)`

**Comment by dmbates** [comment]
I should have been more explicit.  If you fit a linear model using the Iteratively Reweighted Least Squares (IRLS) algorithm it should converge in a single iteration.  The solution can be calculated directly.   Occasionally the convergence criterion is not satisfied at the first iteration due to round-off, it which case it will never be satisfied.

**Comment by nalimilan** [comment]
Good catch. It would be nice to be able to fit a GLM with a normal distribution and identity link, if only for consistency.

**Comment by dmbates** [comment]
@nalimilan Trying to do that gets into tricky issues of assessing convergence.  With that combination nothing changes between iterations because the variance function is a constant and the derivative of the inverse link is a constant.  To me it seems that you will need to incorporate a lot of rules about convergence (maximum iteration count, lack of progress, etc.) to handle this one case unless you evaluate the call to `glm` as the equivalent call to `lm`.  That, of course, would cause the problem of
```julia
fit(GeneralizedLinearModel, ...)
```
returning a `LinearModel` instead.

**Comment by nalimilan** [comment]
I have no idea how hard it would be to implement. Maybe it would be easier to use a specialized method which would fit a `LinearModel` and use its results to fill a `GLM` object?

**Comment by dmbates** [comment]
I think that would be the best approach - to fit the `LinearModel` then make it look like a `GeneralizedLinearModel`.  I know it is not considered good style in Julia to issue warnings but I would consider ways to try to inform the user that they are doing something dumb by fitting a linear model this way.

**Comment by andreasnoack** [comment]
1. Why would a convergence problem show up as a stack overflow?
2. Should it be possible to make the convergence criterion working in the linear case? At least I'd like to know why it would fail.

**Comment by andreasnoack** [comment]
This is not about the convergence. It works just fine. This is about `Union{T,Missing}`. I accidentally created the data frame differently so with
```julia
using GLM, TextParse, DataFrames

data_reg_tp = readcsv("/Users/andreasnoack/Downloads/KORV_Frictions_Reg.csv", has_header=true)
df = DataFrame([data_reg_tp[1][:,i] for i in 1:14])
names!(df, Symbol.(data_reg_tp[2][:]))

regh_u = glm(@formula(L_U ~ K_EQ_CR + K_EQ_LAG + K_STR_CR + K_STR_LAG + REL_P_EQ_LAG + IND_BUS_CYC_LAG + YEAR), 
    df, Normal(), IdentityLink())
```
it works fine (though I also get the deprecation warning). However with
```julia
data_reg = readtable("/Users/andreasnoack/Downloads/KORV_Frictions_Reg.csv", header = true)
```
it fails. The difference between the data frames is
```
julia> eltype.(df.columns)
14-element Array{DataType,1}:
 Float64
 Float64
 Float64
 Float64
 Float64
 Float64
 Float64
 Float64
 Float64
 Float64
 Float64
 Float64
 Float64
 Float64

julia> eltype.(data_reg.columns)
14-element Array{Union,1}:
 Union{Int64, Missings.Missing}
 Union{Float64, Missings.Missing}
 Union{Float64, Missings.Missing}
 Union{Float64, Missings.Missing}
 Union{Float64, Missings.Missing}
 Union{Float64, Missings.Missing}
 Union{Float64, Missings.Missing}
 Union{Float64, Missings.Missing}
 Union{Float64, Missings.Missing}
 Union{Float64, Missings.Missing}
 Union{Float64, Missings.Missing}
 Union{Float64, Missings.Missing}
 Union{Float64, Missings.Missing}
 Union{Float64, Missings.Missing}
```

**Comment by andreasnoack** [comment]
More specifically
```julia
julia> df[:L_U] isa GLM.FPVector
true

julia> data_reg[:L_U] isa GLM.FPVector
false
```
so it can't enter https://github.com/JuliaStats/GLM.jl/blob/287c668acd592c05ec90140dfa70a2887045b4f3/src/glmfit.jl#L258-L264 which is restricted to `y::FPVector`. I'm not sure what the right fix here is but it would be great to handle it in a `Missing` agnostic way, i.e. without using `Missings` explicitly anywhere in `GLM`.

**Comment by andreasnoack** [comment]
Sorry for continuing my monologue here. We probably need some kind of deunionization somewhere in the pipeline to avoid dispatch to slow methods, i.e. `Vector{Union{Float64,Missing}}` can't to go BLAS so with unions we risk things like
```julia
julia> X = randn(10000,10);

julia> y1 = randn(10000);

julia> y2 = Vector{Union{Float64,Missing}}(y1);

julia> @time X'y1;
  0.000126 seconds (5 allocations: 320 bytes)

julia> @time X'y2;
  0.005866 seconds (500.12 k allocations: 7.631 MiB)
```

**Comment by kleinschmidt** [comment]
The right place for this is in `StatsModels.model_response`.  I've opened an issue there.

**Comment by andreasnoack** [comment]
Fixed by c709ec130d420b776b8b246244e6c93b906bff0a
